crane_plus_examples/src/pick_and_place_tf.cpp

```diff
@@ -0,0 +1,228 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Reference:
+// https://github.com/ros-planning/moveit2_tutorials/blob
+// /a547cf49ff7d1fe16a93dfe020c6027bcb035b51/doc/move_group_interface
+// /src/move_group_interface_tutorial.cpp
+// https://docs.ros.org/en/humble/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Listener-Cpp.html
+
+#include <chrono>
+#include <cmath>
+#include <memory>
+
+#include "angles/angles.h"
+#include "geometry_msgs/msg/pose.hpp"
+#include "geometry_msgs/msg/quaternion.hpp"
+#include "geometry_msgs/msg/transform_stamped.hpp"
+#include "geometry_msgs/msg/twist.hpp"
+#include "moveit/move_group_interface/move_group_interface.h"
+#include "rclcpp/rclcpp.hpp"
+#include "tf2_geometry_msgs/tf2_geometry_msgs.h"
+#include "tf2/convert.h"
+#include "tf2/exceptions.h"
+#include "tf2_ros/transform_listener.h"
+#include "tf2_ros/buffer.h"
+#include "std_msgs/msg/string.hpp"
+using namespace std::chrono_literals;
+using MoveGroupInterface = moveit::planning_interface::MoveGroupInterface;
+
+class PickAndPlaceTf : public rclcpp::Node
+{
+public:
+  PickAndPlaceTf(
+    rclcpp::Node::SharedPtr move_group_arm_node,
+    rclcpp::Node::SharedPtr move_group_gripper_node)
+  : Node("pick_and_place_tf_node")
+  {
+    using namespace std::placeholders;
+    move_group_arm_ = std::make_shared<MoveGroupInterface>(move_group_arm_node, "arm");
+    move_group_arm_->setMaxVelocityScalingFactor(1.0);
+    move_group_arm_->setMaxAccelerationScalingFactor(1.0);
+
+    move_group_gripper_ = std::make_shared<MoveGroupInterface>(move_group_gripper_node, "gripper");
+    move_group_gripper_->setMaxVelocityScalingFactor(1.0);
+    move_group_gripper_->setMaxAccelerationScalingFactor(1.0);
+
+    // SRDFに定義されている"home"の姿勢にする
+    move_group_arm_->setNamedTarget("home");
+    move_group_arm_->move();
+
+    // 可動範囲を制限する
+    moveit_msgs::msg::Constraints constraints;
+    constraints.name = "arm_constraints";
+
+    moveit_msgs::msg::JointConstraint joint_constraint;
+    joint_constraint.joint_name = "crane_plus_joint1";
+    joint_constraint.position = 0.0;
+    joint_constraint.tolerance_above = angles::from_degrees(120);
+    joint_constraint.tolerance_below = angles::from_degrees(120);
+    joint_constraint.weight = 1.0;
+    constraints.joint_constraints.push_back(joint_constraint);
+
+    move_group_arm_->setPathConstraints(constraints);
+
+    // 待機姿勢
+    control_arm(0.0, 0.0, 0.17, 0, 90, 0);
+
+    tf_buffer_ =
+      std::make_unique<tf2_ros::Buffer>(this->get_clock());
+    tf_listener_ =
+      std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+
+    timer_ = this->create_wall_timer(
+      500ms, std::bind(&PickAndPlaceTf::on_timer, this));
+  }
+
+private:
+  void on_timer()
+  {
+    // 把持物体の位置を取得
+    geometry_msgs::msg::TransformStamped tf_msg;
+
+    try {
+      tf_msg = tf_buffer_->lookupTransform(
+        "base_link", "target",
```

Comment:
ARマーカサンプルと互換性をもたせるためframe名を変更
```suggestion
        "base_link", "target_0",
```

Reply:
targetをtarget_0に変更しました。